[gnus] Make the gnus variables user configurable

duianto opened the pull request:
The gnus variables were set in the `gnus/init-gnus`
functions `:config` section, that made them override
any user configurations.

Moving the variable declarations to the `:init`
section made them user configurable.

Fixes #13268

#### Note
The diff doesn't match the description.

The variables were moved up a about 15 lines, but the diff shows the configurations that were above the variables being moved down instead.

---

#### Previous description
(it still applies, but the change described above applies to all the variables that the `gnus` layer configures)

Made `gnus-summary-thread-gathering-function` user changeable
as a layer variable.

---

With the following `gnus` layer variable:
```elisp
     (gnus :variables
           gnus-summary-thread-gathering-function 'gnus-gather-threads-by-subject)
```

### Before this PR

#### Before opening a gnus buffer `SPC a g g`:
>gnus-summary-thread-gathering-function’s value is
‘gnus-gather-threads-by-subject’

#### After opening a gnus buffer:
>gnus-summary-thread-gathering-function is a variable defined in ‘gnus-sum.el’.
Its value is ‘gnus-gather-threads-by-references’
Original value was 
gnus-gather-threads-by-subject

### With this PR

#### Before opening a gnus buffer:
>gnus-summary-thread-gathering-function is a variable defined in ‘config.el’.
Its value is ‘gnus-gather-threads-by-subject’

#### After opening a gnus buffer:
>gnus-summary-thread-gathering-function is a variable defined in ‘gnus-sum.el’.
Its value is ‘gnus-gather-threads-by-subject’

duianto commented:
I suspect that we're going to run into the same issue when someone want's to configure any of the other variables that the layer sets in the `gnus/init-gnus` functions `:config` section:
https://github.com/syl20bnr/spacemacs/blob/09a9273e225c84ed9c7b6c5c4f5b50ab721ee57c/layers/%2Bemail/gnus/packages.el#L59-L79

Would it be enough to move their definitions to the `:init` section?
Or should all of them be redefined as layer variables?

smile13241324 commented:
I would avoid duplicating more variables from the gnus package in the layer as long as the layer does not add any additional meaning to the variable. Lets just move them into the init function and mention in the README that these variables can be configured in the user-config or in the layer declaration.

duianto commented:
The diff doesn't show the variables being moved up. Instead it shows the configurations that were above the variables, being moved down.

I'm not sure how to fix that.

---

I moved down the lines at the beginning of the `:init` progn, so that everything in the init section are indented under the progn:

Before:
```elisp
    (progn (spacemacs/declare-prefix "ag" "gnus" "Gnus newsreader")
           (spacemacs/set-leader-keys
             "agg" 'gnus
             "ags" 'gnus-slave
             "agu" 'gnus-unplugged
...
```
After:
```elisp
    (progn
      (spacemacs/declare-prefix "ag" "gnus" "Gnus newsreader")
      (spacemacs/set-leader-keys
        "agg" 'gnus
        "ags" 'gnus-slave
        "agu" 'gnus-unplugged
...
```

---

For the readme.org mentions of the variables, I added a source block with a straight copy of the variables and their values.

I added a comment before the variable definitions in `packages.el`, that any change to the variables also should be updated in readme.org.

smile13241324 commented:
Looks good, Microsoft must have fixed the diff issue already I see that the
vars have been moved to the init section.

On Sun, 1 Mar 2020, 23:23 duianto, <notifications@github.com> wrote:

> The diff doesn't show the variables being moved up. Instead it shows the
> configurations that were above the variables, being moved down.
>
> I'm not sure how to fix that.
> ------------------------------
>
> I moved down the lines at the beginning of the :init progn, so that
> everything in the init section are indented under the progn:
>
> Before:
>
>     (progn (spacemacs/declare-prefix "ag" "gnus" "Gnus newsreader")
>            (spacemacs/set-leader-keys
>              "agg" 'gnus
>              "ags" 'gnus-slave
>              "agu" 'gnus-unplugged
> ...
>
> After:
>
>     (progn
>       (spacemacs/declare-prefix "ag" "gnus" "Gnus newsreader")
>       (spacemacs/set-leader-keys
>         "agg" 'gnus
>         "ags" 'gnus-slave
>         "agu" 'gnus-unplugged
> ...
>
> ------------------------------
>
> For the readme.org mentions of the variables, I added a source block with
> a straight copy of the variables and their values.
>
> I added a comment before the variable definitions in packages.el, that
> any change to the variables also should be updated in readme.org.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/13310?email_source=notifications&email_token=AEKCTAFYYH3CS2FWWIO4MLDRFLN43A5CNFSM4K4NW4J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENNMYSI#issuecomment-593153097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKCTAB4QBHBKYYXVJ4XQATRFLN43ANCNFSM4K4NW4JQ>
> .
>

duianto commented:
Nothing seems to have changed for me: https://github.com/syl20bnr/spacemacs/pull/13310/files
it still shows the `:config` line and the following 4 paragraphs being moved below the setq-default assigned variables.

But I might have found a possible solution:
If the variables are moved and indented to match the current entries in the `:init` section, to the right of the `(progn`. Then the variables are shown as being moved.
I'll include a second commit that corrects the `:init` section to be indented below the `(progn`.

duianto commented:
Nothing changed in the combined diff:
https://github.com/syl20bnr/spacemacs/pull/13310/files

But individually each commits diff, now matches their descriptions:
[gnus] Make the gnus variables user configurable https://github.com/syl20bnr/spacemacs/pull/13310/commits/f2c61d0e8f0525f00905d025680ca08fadbe6fd9
[gnus] Fix gnus/init-gnus :init section indentation https://github.com/syl20bnr/spacemacs/pull/13310/commits/82ecb39834dcd75a89e6b51d8eaf9f584aeb99b1

smile13241324 commented:
Merged

duianto commented:
Yes, very strange.

The GitHub diff looks the same for me in:
- Firefox, Chrome, Brave and Edge (based on chromium) in Windows 1903
- Firefox and Chromium in PopOS 19.10

The `set-default` line looks unmodified in:
- This PRs current combined diff: https://github.com/syl20bnr/spacemacs/pull/13310/files
- In the commit from before the PR was split into two commits: https://github.com/syl20bnr/spacemacs/commit/f1703a3d6bece9ec8ac2aa7ea834a3064912b457
- And on the Spacemacs commit from before this PRs two commits:
Bring ipython-notebook up to date. e0aee1554454aff8183076b442b2903e5fa072c1

  When the `setq-default` variables are moved from the `:config` to the `:init` section, then the issue is also observed in both Spacemacs and when calling `git diff` from the Windows command prompt:

![cmd_2020-03-05_09-49-36](https://user-images.githubusercontent.com/13420573/75964116-ee229200-5ec6-11ea-8751-b0612b448704.png)

No response required, this just documents that I'm still seeing the issue in Firefox.
